### PR TITLE
Publish SNS event on new alert

### LINKED ISSUE #198: Feat - Implement SNS Client as Alert Option
### Problem
Allowing strictly Slack and PagerDuty as a way for users to receive alerts is a limiting factor for alert consumption/ingestion.

### Solution
Implementing a configurable, optional SNS client for publishing events will enable users to receive emails and SMS messages, and create custom automation pipelines to ingest alerts upon heuristic triggers.

### PR DESCRIPTION
**Pull request by adrain-cb**
# Pull Request Template

## Fixes Issue

Fixes #198

### CODE REVIEW

File: internal/alert/manager.go [hunk 1]
```diff
@@ -39,28 +40,30 @@ type alertManager struct {
 	interpolator *Interpolator
 	cdHandler    CoolDownHandler
 	cm           RoutingDirectory
+	sns          client.SNSClient
```
[the remark this answers is not in the document]
Comment (adrain-cb) [reply internal/alert/manager.go:43]: I could add the client to the routing directory instead

[the remark this answers is not in the document]
Comment (adrain-cb) [reply internal/alert/manager.go]: [e7a88d0](https://github.com/base-org/pessimism/pull/200/commits/e7a88d07e094325f966862bcbf850c9b86fc54c7)

File: scripts/localstack-setup.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
```
Comment (epociask) [inline scripts/localstack-setup.sh:1]: maybe update the script name to better reflect that its unique to testing

File: e2e/alerting_test.go [hunk 3]
```diff
@@ -159,15 +172,23 @@ func TestCoolDown(t *testing.T) {
 		return height != nil && height.Uint64() > receipt.BlockNumber.Uint64(), nil
 	}))
 
+	time.Sleep(1 * time.Second)
```
Comment (epociask) [inline e2e/alerting_test.go:175]: Why are we sleeping here? The requirement loop should ensure that pessimism is synced with the latest block height

File: config.env.template [hunk 4]
```diff
@@ -32,6 +32,8 @@ P0_PAGERDUTY_ALERT_EVENTS_URL=
 P1_PAGERDUTY_INTEGRATION_KEY=
 P1_PAGERDUTY_ALERT_EVENTS_URL=
 
+SNS_TOPIC_ARN=
```
Comment (epociask) [inline config.env.template:35]: I feel like we should update the documentation to reflect that an SNS routing feature is now supported

File: internal/alert/manager.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ type Config struct {
 	RoutingCfgPath          string
 	PagerdutyAlertEventsURL string
 	RoutingParams           *core.AlertRoutingParams
+	SNSConfig               *client.SNSConfig
```
Comment (epociask) [inline internal/alert/manager.go:30]: Can we add a comment detailing why this alerting resource isn't part of the `AlertRoutingParams`?

File: internal/alert/manager.go [hunk 6]
```diff
@@ -135,13 +136,36 @@ func (am *alertManager) handlePagerDutyPost(alert core.Alert) error {
...
 }
 
+func (am *alertManager) handleSNSPublish(alert core.Alert, policy *core.AlertPolicy) error {
+	event := &client.AlertEventTrigger{
+		Message:  am.interpolator.SlackMessage(alert, policy.Msg),
+		DedupKey: alert.PathID,
+		Severity: alert.Sev,
+	}
+
+	cli := am.cm.GetSNSClient()
```
Comment (epociask) [inline internal/alert/manager.go:153]: Knit - cli implies command line interface but is being used to represent client - maybe just change var to `client` or `c`

File: internal/alert/manager.go [hunk 7]
```diff
@@ -135,13 +136,36 @@ func (am *alertManager) handlePagerDutyPost(alert core.Alert) error {
...
+		am.logger.Debug("Successfully posted to PagerDuty", zap.Any("resp", resp))
 		am.metrics.RecordAlertGenerated(alert, core.PagerDuty, pdc.GetName())
 	}
 
 	return nil
 }
 
+func (am *alertManager) handleSNSPublish(alert core.Alert, policy *core.AlertPolicy) error {
+	event := &client.AlertEventTrigger{
+		Message:  am.interpolator.SlackMessage(alert, policy.Msg),
```
Comment (epociask) [inline internal/alert/manager.go:148]: The messages we post to slack are unstructured and contain emojis. Perhaps a JSON structured payload would be better fit for an sns subscriber to reason with?

File: internal/alert/routing.go [hunk 8]
```diff
@@ -22,6 +24,7 @@ type RoutingDirectory interface {
 type routingDirectory struct {
 	pagerDutyClients map[core.Severity][]client.PagerDutyClient
 	slackClients     map[core.Severity][]client.SlackClient
+	snsClient        client.SNSClient
```
Comment (epociask) [inline internal/alert/routing.go:27]: Lets comment why this deviates from its above field counterparts

File: internal/client/sns.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+func NewSNSClient(cfg *SNSConfig, name string) SNSClient {
+	if cfg.TopicArn == "" {
+		logging.NoContext().Warn("No SNS topic ARN provided")
+	}
+
+	// Initialize a session that the SDK will use to load configuration,
+	// credentials, and region. AWS_REGION, AWS_SECRET_ACCESS_KEY, and AWS_ACCESS_KEY_ID should be set in the
+	// environment's runtime
+	sess, err := session.NewSession(&aws.Config{
+		Endpoint: aws.String(cfg.Endpoint),
```
Comment (epociask) [inline internal/client/sns.go:44]: What would happen if this session were to arbitrarily crash? Would the client do a reset or would all subsequent publishing fail to work until the application is reset?

File: internal/metrics/metrics.go [hunk 10]
```diff
@@ -247,8 +247,9 @@ func (m *Metrics) RecordAlertGenerated(alert core.Alert, dest core.AlertDestinat
 	net := alert.PathID.Network().String()
 	h := alert.HT.String()
 	sev := alert.Sev.String()
+	path := alert.PathID.String()
```
Comment (epociask) [inline internal/metrics/metrics.go:250]: Knit:
```suggestion
	id := alert.PathID.String()
```

File: docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,15 @@
+version: "3.8"
+
+services:
+  localstack:
+    container_name: "${LOCALSTACK_DOCKER_NAME:-localstack-main}"
+    image: localstack/localstack
```
Comment (epociask) [inline docker-compose.yml:6]: Does this lookup the most recent image? If so it may be a good idea to do commit pinning to reduce our security attack surface.

File: e2e/heuristic_test.go [hunk 12]
```diff
@@ -31,7 +31,7 @@ import (
 // balance enforcement heuristic session on L2 network.
 func TestBalanceEnforcement(t *testing.T) {
 
-	ts := e2e.CreateSysTestSuite(t)
+	ts := e2e.CreateSysTestSuite(t, "")
```
Comment (epociask) [inline e2e/heuristic_test.go:34]: Don't love the idea of extending this constructor with a single string value - is there not a way to encapsulate this?

File: e2e/setup.go [hunk 13]
```diff
@@ -112,11 +115,22 @@ func CreateSysTestSuite(t *testing.T) *SysTestSuite {
...
 
+	if err := os.Setenv("AWS_REGION", "us-east-1"); err != nil { //nolint:tenv // Cannot use t.SetEnv here
+		t.Fatal(err)
+	}
+	if err := os.Setenv("AWS_SECRET_ACCESS_KEY", "test"); err != nil { //nolint:tenv // Cannot t.Setenv here
+		t.Fatal(err)
+	}
+	if err := os.Setenv("AWS_ACCESS_KEY_ID", "test"); err != nil { //nolint:tenv // Cannot use t.SetEnv here
+		t.Fatal(err)
+	}
```
Comment (epociask) [inline e2e/setup.go:126]:
```suggestion
func setAwsVars() {
	awsEnvVariables := map[string]string{
		"AWS_REGION":            "us-east-1",
		"AWS_SECRET_ACCESS_KEY": "test",
		"AWS_ACCESS_KEY_ID":     "test",
	}

	for key, value := range awsEnvVariables {
		if err := os.Setenv(key, value); err != nil {
			t.Fatalf("Error setting %s environment variable: %s", key, err)
		}
	}
}
```

**Review by epociask — COMMENTED** [review]
This looks good, just have some comments 🔥

File: e2e/alerting_test.go (hunk 3 above)
Comment (adrain-cb) [reply e2e/alerting_test.go:175]: Yeah was testing something and forgot to delete this

File: e2e/heuristic_test.go (hunk 12 above)
Comment (adrain-cb) [reply e2e/heuristic_test.go:34]: I agree, but I couldn't think of a better way to solve this. Open to any ideas here

File: internal/client/sns.go (hunk 9 above)
Comment (adrain-cb) [reply internal/client/sns.go:44]: Hmm that's a good question. From the [docs](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/) I've read, I haven't been able to find an answer to this. Maybe we should expect that the session could arbitrarily crash just in the case that it does? If this is the case, I can move the session creation to the function that handles publishing the message, but I'm not sure that's the right way to handle this. Open to ideas here

File: e2e/setup.go (hunk 13 above)
Comment (adrain-cb) [reply e2e/setup.go]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: docker-compose.yml (hunk 11 above)
Comment (adrain-cb) [reply docker-compose.yml]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: internal/metrics/metrics.go (hunk 10 above)
Comment (adrain-cb) [reply internal/metrics/metrics.go]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: internal/alert/routing.go (hunk 8 above)
Comment (adrain-cb) [reply internal/alert/routing.go:28]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: internal/alert/manager.go (hunk 7 above)
Comment (adrain-cb) [reply internal/alert/manager.go]: Added a quick unit test to document how the message is marshalled/unmarshalled since aws requires some odd nesting

File: internal/alert/manager.go (hunk 6 above)
Comment (adrain-cb) [reply internal/alert/manager.go]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: internal/alert/manager.go (hunk 5 above)
Comment (adrain-cb) [reply internal/alert/manager.go:31]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: e2e/alerting_test.go (hunk 3 above)
Comment (adrain-cb) [reply e2e/alerting_test.go]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: scripts/localstack-setup.sh (hunk 2 above)
Comment (adrain-cb) [reply scripts/localstack-setup.sh]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: e2e/setup.go [hunk 14]
```diff
@@ -186,6 +203,39 @@ func DefaultTestConfig() *config.Config {
 	}
 }
 
+func GetMessages(endpoint string, queueName string) (*sqs.ReceiveMessageOutput, error) {
```
[the remark this answers is not in the document]
Comment (adrain-cb) [reply e2e/setup.go]: [74f2414](https://github.com/base-org/pessimism/pull/200/commits/74f241413764e3b9caf17a604ffe27f2bc28480c)

File: config.env.template (hunk 4 above)
Comment (adrain-cb) [reply config.env.template:35]: [655294b](https://github.com/base-org/pessimism/pull/200/commits/655294b1b10c0d11ed7a34b5f42d87540a31c7e6)

File: e2e/alerting_test.go [hunk 15]
```diff
@@ -73,6 +80,12 @@ func TestMultiDirectiveRouting(t *testing.T) {
 		return height != nil && height.Uint64() > receipt.BlockNumber.Uint64(), nil
 	}))
 
+	sqsMessages, err := e2e.GetSNSMessages(ts.AppCfg.AlertConfig.SNSConfig.Endpoint, "multi-directive-test-queue")
```
Comment (epociask) [inline e2e/alerting_test.go:83]:
```suggestion
	snsMessages, err := e2e.GetSNSMessages(ts.AppCfg.AlertConfig.SNSConfig.Endpoint, "multi-directive-test-queue")
```

File: e2e/alerting_test.go [hunk 16]
```diff
@@ -162,6 +175,11 @@ func TestCoolDown(t *testing.T) {
 	// Check that the balance enforcement was triggered using the mocked server cache.
 	posts := ts.TestSlackSvr.SlackAlerts()
 
+	sqsMessages, err := e2e.GetSNSMessages(ts.AppCfg.AlertConfig.SNSConfig.Endpoint, "alert-cooldown-test-queue")
```
Comment (epociask) [inline e2e/alerting_test.go:178]:
```suggestion
	snsMessages, err := e2e.GetSNSMessages(ts.AppCfg.AlertConfig.SNSConfig.Endpoint, "alert-cooldown-test-queue")
```

File: internal/client/sns.go (hunk 9 above)
Comment (epociask) [reply internal/client/sns.go:62]: Not super pressing but definitely an edge case to lookout for in the wild - lets just add a comment explaining that this could happen

**Review by epociask — COMMENTED** [review]
Just a few knits - once those are addressed it should be good to go 💯

File: e2e/alerting_test.go (hunk 15 above)
Comment (adrain-cb) [reply e2e/alerting_test.go]: [bc97527](https://github.com/base-org/pessimism/pull/200/commits/bc975275398280809302d537705d883a7a973c39)

File: internal/client/sns.go (hunk 9 above)
Comment (adrain-cb) [reply internal/client/sns.go:63]: [bc97527](https://github.com/base-org/pessimism/pull/200/commits/bc975275398280809302d537705d883a7a973c39)

**Review by epociask — DISMISSED** [review]
LGTM 🧨

**Comment by cb-heimdall** [comment]
Review Error for epociask @ 2024-02-09 22:12:01 UTC
User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)

File: e2e/alerting_test.go [hunk 17]
```diff
@@ -17,11 +18,17 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+const (
+	MultiDirectiveTopicArn = "arn:aws:sns:us-east-1:000000000000:multi-directive-test-topic"
```
Comment (megha31-cb) [inline e2e/alerting_test.go:22]: Can we add a comment stating these are localstack specific values to simulate aws arns?

File: e2e/alerting_test.go (hunk 16 above)
Comment (megha31-cb) [inline e2e/alerting_test.go:178]: Shouldn't we test publish messages to SNS?

File: internal/client/sns.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+func NewSNSClient(cfg *SNSConfig, name string) SNSClient {
+	if cfg.TopicArn == "" {
+		logging.NoContext().Warn("No SNS topic ARN provided")
+	}
+
+	// Initialize a session that the SDK will use to load configuration,
+	// credentials, and region. AWS_REGION, AWS_SECRET_ACCESS_KEY, and AWS_ACCESS_KEY_ID should be set in the
+	// environment's runtime
+	// Note: If session is to arbitrarily crash, there is a possibility that message publishing will fail
+	sess, err := session.NewSession(&aws.Config{
```
Comment (megha31-cb) [inline internal/client/sns.go:62]: NIT: You can create session as a singleton object

File: e2e/alerting_test.go (hunk 16 above)
Comment (adrain-cb) [reply e2e/alerting_test.go:178]: This gets messages that were published by Pessimism, so this test ensures a message was published.
